File: bindings/node-open/src/panchang.js
'use strict';

const { addon } = require('./native');
const { checkStatus } = require('./errors');

function bhavaSystemCount() {
  return addon.bhavaSystemCount();
}

function computeRiseSet(engine, eop, location, config, eventCode, jdTdbApprox, lsk) {
  const r = addon.computeRiseSet(engine._handle, eop._handle, location, config, eventCode, jdTdbApprox, lsk._handle);
  checkStatus('compute_rise_set', r.status);
  return r.result;
}

function computeAllEvents(engine, eop, location, config, jdTdbApprox, lsk) {
  const r = addon.computeAllEvents(engine._handle, eop._handle, location, config, jdTdbApprox, lsk._handle);
  checkStatus('compute_all_events', r.status);
  return r.results;
}

function computeRiseSetUtc(engine, eop, lsk, location, eventCode, utc, config) {
  const r = addon.computeRiseSetUtc(engine._handle, eop._handle, lsk._handle, location, eventCode, utc, config);
  checkStatus('compute_rise_set_utc', r.status);
  return r.result;
}

function computeAllEventsUtc(engine, eop, lsk, location, utc, config) {
  const r = addon.computeAllEventsUtc(engine._handle, eop._handle, lsk._handle, location, utc, config);
  checkStatus('compute_all_events_utc', r.status);
  return r.results;
}

function computeBhavas(engine, eop, location, lsk, jdTdb, config) {
  const r = addon.computeBhavas(engine._handle, eop._handle, location, lsk._handle, jdTdb, config);
  checkStatus('compute_bhavas', r.status);
  return r.bhava;
}

function computeBhavasUtc(engine, eop, lsk, location, utc, config) {
  const r = addon.computeBhavasUtc(engine._handle, eop._handle, lsk._handle, location, utc, config);
  checkStatus('compute_bhavas_utc', r.status);
  return r.bhava;
}

function lagnaDeg(lsk, eop, location, jdTdb, config) {
  const r = config == null
    ? addon.lagnaDeg(lsk._handle, eop._handle, location, jdTdb)
    : addon.lagnaDeg(lsk._handle, eop._handle, location, jdTdb, config);
  checkStatus('lagna_deg', r.status);
  return r.degrees;
}

function mcDeg(lsk, eop, location, jdTdb, config) {
  const r = config == null
    ? addon.mcDeg(lsk._handle, eop._handle, location, jdTdb)
    : addon.mcDeg(lsk._handle, eop._handle, location, jdTdb, config);
  checkStatus('mc_deg', r.status);
  return r.degrees;
}

function ramcDeg(lsk, eop, location, jdTdb) {
  const r = addon.ramcDeg(lsk._handle, eop._handle, location, jdTdb);
  checkStatus('ramc_deg', r.status);
  return r.degrees;
}

function lagnaDegUtc(lsk, eop, location, utc, config) {
  const r = config == null
    ? addon.lagnaDegUtc(lsk._handle, eop._handle, location, utc)
    : addon.lagnaDegUtc(lsk._handle, eop._handle, location, utc, config);
  checkStatus('lagna_deg_utc', r.status);
  return r.degrees;
}

function mcDegUtc(lsk, eop, location, utc, config) {
  const r = config == null
    ? addon.mcDegUtc(lsk._handle, eop._handle, location, utc)
    : addon.mcDegUtc(lsk._handle, eop._handle, location, utc, config);
  checkStatus('mc_deg_utc', r.status);
  return r.degrees;
}

function ramcDegUtc(lsk, eop, location, utc) {
  const r = addon.ramcDegUtc(lsk._handle, eop._handle, location, utc);
  checkStatus('ramc_deg_utc', r.status);
  return r.degrees;
}

function riseSetResultToUtc(lsk, riseSetResult) {
  const r = addon.riseSetResultToUtc(lsk._handle, riseSetResult);
  checkStatus('riseset_result_to_utc', r.status);
  return r.utc;
}

function tithiForDate(engine, utc) {
  const r = addon.tithiForDate(engine._handle, utc);
  checkStatus('tithi_for_date', r.status);
  return r.tithi;
}

function karanaForDate(engine, utc) {
  const r = addon.karanaForDate(engine._handle, utc);
  checkStatus('karana_for_date', r.status);
  return r.karana;
}

function yogaForDate(engine, utc, config = addon.sankrantiConfigDefault()) {
  const r = addon.yogaForDate(engine._handle, utc, config);
  checkStatus('yoga_for_date', r.status);
  return r.yoga;
}

function nakshatraForDate(engine, utc, config = addon.sankrantiConfigDefault()) {
  const r = addon.nakshatraForDate(engine._handle, utc, config);
  checkStatus('nakshatra_for_date', r.status);
  return r.nakshatra;
}

function vaarForDate(engine, eop, utc, location, config = addon.riseSetConfigDefault()) {
  const r = addon.vaarForDate(engine._handle, eop._handle, utc, location, config);
  checkStatus('vaar_for_date', r.status);
  return r.vaar;
}

function horaForDate(engine, eop, utc, location, config = addon.riseSetConfigDefault()) {
  const r = addon.horaForDate(engine._handle, eop._handle, utc, location, config);
  checkStatus('hora_for_date', r.status);
  return r.hora;
}

function ghatikaForDate(engine, eop, utc, location, config = addon.riseSetConfigDefault()) {
  const r = addon.ghatikaForDate(engine._handle, eop._handle, utc, location, config);
  checkStatus('ghatika_for_date', r.status);
  return r.ghatika;
}

function masaForDate(engine, utc, config = addon.sankrantiConfigDefault()) {
  const r = addon.masaForDate(engine._handle, utc, config);
  checkStatus('masa_for_date', r.status);
  return r.masa;
}

function ayanaForDate(engine, utc, config = addon.sankrantiConfigDefault()) {
  const r = addon.ayanaForDate(engine._handle, utc, config);
  checkStatus('ayana_for_date', r.status);
  return r.ayana;
}

function varshaForDate(engine, utc, config = addon.sankrantiConfigDefault()) {
  const r = addon.varshaForDate(engine._handle, utc, config);
  checkStatus('varsha_for_date', r.status);
  return r.varsha;
}

function panchangComputeEx(engine, eop, lsk, request) {
  const r = addon.panchangComputeEx(engine._handle, eop._handle, lsk._handle, request);
  checkStatus('panchang_compute_ex', r.status);
  return r.result;
}

module.exports = {
  bhavaSystemCount,
  computeRiseSet,
  computeAllEvents,
  computeRiseSetUtc,
  computeAllEventsUtc,
  computeBhavas,
  computeBhavasUtc,
  lagnaDeg,
  mcDeg,
  ramcDeg,
  lagnaDegUtc,
  mcDegUtc,
  ramcDegUtc,
  riseSetResultToUtc,
  tithiForDate,
  karanaForDate,
  yogaForDate,
  nakshatraForDate,
  vaarForDate,
  horaForDate,
  ghatikaForDate,
  masaForDate,
  ayanaForDate,
  varshaForDate,
  panchangComputeEx,
};

File: bindings/go-open/dhruv/panchang.go
package dhruv

import "ctara-dhruv-core/bindings/go-open/internal/cabi"

func RiseSetConfigDefault() RiseSetConfig { return cabi.RiseSetConfigDefault() }
func BhavaConfigDefault() BhavaConfig     { return cabi.BhavaConfigDefault() }

func BhavaSystemCount() uint32 { return cabi.BhavaSystemCount() }

func (e *Engine) ComputeRiseSet(ep *EOP, loc GeoLocation, cfg RiseSetConfig, eventCode int32, jdTdbApprox float64, lsk *LSK) (RiseSetResult, error) {
	out, st := cabi.ComputeRiseSet(e.h, ep.h, loc, cfg, eventCode, jdTdbApprox, lsk.h)
	return out, statusErr("compute_rise_set", st)
}

func (e *Engine) ComputeAllEvents(ep *EOP, loc GeoLocation, cfg RiseSetConfig, jdTdbApprox float64, lsk *LSK) ([8]RiseSetResult, error) {
	out, st := cabi.ComputeAllEvents(e.h, ep.h, loc, cfg, jdTdbApprox, lsk.h)
	return out, statusErr("compute_all_events", st)
}

func (e *Engine) ComputeRiseSetUTC(ep *EOP, lsk *LSK, loc GeoLocation, eventCode int32, utc UtcTime, cfg RiseSetConfig) (RiseSetResultUTC, error) {
	out, st := cabi.ComputeRiseSetUTC(e.h, ep.h, lsk.h, loc, eventCode, utc, cfg)
	return out, statusErr("compute_rise_set_utc", st)
}

func (e *Engine) ComputeAllEventsUTC(ep *EOP, lsk *LSK, loc GeoLocation, utc UtcTime, cfg RiseSetConfig) ([8]RiseSetResultUTC, error) {
	out, st := cabi.ComputeAllEventsUTC(e.h, ep.h, lsk.h, loc, utc, cfg)
	return out, statusErr("compute_all_events_utc", st)
}

func (e *Engine) ComputeBhavas(ep *EOP, loc GeoLocation, lsk *LSK, jdTdb float64, cfg BhavaConfig) (BhavaResult, error) {
	out, st := cabi.ComputeBhavas(e.h, ep.h, loc, lsk.h, jdTdb, cfg)
	return out, statusErr("compute_bhavas", st)
}

func (e *Engine) ComputeBhavasUTC(ep *EOP, lsk *LSK, loc GeoLocation, utc UtcTime, cfg BhavaConfig) (BhavaResult, error) {
	out, st := cabi.ComputeBhavasUTC(e.h, ep.h, lsk.h, loc, utc, cfg)
	return out, statusErr("compute_bhavas_utc", st)
}

func (e *Engine) LagnaDeg(lsk *LSK, ep *EOP, loc GeoLocation, jdTdb float64) (float64, error) {
	out, st := cabi.LagnaDeg(lsk.h, ep.h, loc, jdTdb)
	return out, statusErr("lagna_deg", st)
}

func (e *Engine) LagnaDegWithConfig(lsk *LSK, ep *EOP, loc GeoLocation, jdTdb float64, cfg BhavaConfig) (float64, error) {
	out, st := cabi.LagnaDegWithConfig(lsk.h, ep.h, loc, jdTdb, cfg)
	return out, statusErr("lagna_deg_with_config", st)
}

func (e *Engine) MCDeg(lsk *LSK, ep *EOP, loc GeoLocation, jdTdb float64) (float64, error) {
	out, st := cabi.MCDeg(lsk.h, ep.h, loc, jdTdb)
	return out, statusErr("mc_deg", st)
}

func (e *Engine) MCDegWithConfig(lsk *LSK, ep *EOP, loc GeoLocation, jdTdb float64, cfg BhavaConfig) (float64, error) {
	out, st := cabi.MCDegWithConfig(lsk.h, ep.h, loc, jdTdb, cfg)
	return out, statusErr("mc_deg_with_config", st)
}

func (e *Engine) RAMCDeg(lsk *LSK, ep *EOP, loc GeoLocation, jdTdb float64) (float64, error) {
	out, st := cabi.RAMCDeg(lsk.h, ep.h, loc, jdTdb)
	return out, statusErr("ramc_deg", st)
}

func (e *Engine) LagnaDegUTC(lsk *LSK, ep *EOP, loc GeoLocation, utc UtcTime) (float64, error) {
	out, st := cabi.LagnaDegUTC(lsk.h, ep.h, loc, utc)
	return out, statusErr("lagna_deg_utc", st)
}

func (e *Engine) LagnaDegUTCWithConfig(lsk *LSK, ep *EOP, loc GeoLocation, utc UtcTime, cfg BhavaConfig) (float64, error) {
	out, st := cabi.LagnaDegUTCWithConfig(lsk.h, ep.h, loc, utc, cfg)
	return out, statusErr("lagna_deg_utc_with_config", st)
}

func (e *Engine) MCDegUTC(lsk *LSK, ep *EOP, loc GeoLocation, utc UtcTime) (float64, error) {
	out, st := cabi.MCDegUTC(lsk.h, ep.h, loc, utc)
	return out, statusErr("mc_deg_utc", st)
}

func (e *Engine) MCDegUTCWithConfig(lsk *LSK, ep *EOP, loc GeoLocation, utc UtcTime, cfg BhavaConfig) (float64, error) {
	out, st := cabi.MCDegUTCWithConfig(lsk.h, ep.h, loc, utc, cfg)
	return out, statusErr("mc_deg_utc_with_config", st)
}

func (e *Engine) RAMCDegUTC(lsk *LSK, ep *EOP, loc GeoLocation, utc UtcTime) (float64, error) {
	out, st := cabi.RAMCDegUTC(lsk.h, ep.h, loc, utc)
	return out, statusErr("ramc_deg_utc", st)
}

func (e *Engine) TithiForDate(utc UtcTime) (TithiInfo, error) {
	out, st := cabi.TithiForDate(e.h, utc)
	return out, statusErr("tithi_for_date", st)
}

func (e *Engine) KaranaForDate(utc UtcTime) (KaranaInfo, error) {
	out, st := cabi.KaranaForDate(e.h, utc)
	return out, statusErr("karana_for_date", st)
}

func (e *Engine) YogaForDate(utc UtcTime, cfg SankrantiConfig) (YogaInfo, error) {
	out, st := cabi.YogaForDate(e.h, utc, cfg)
	return out, statusErr("yoga_for_date", st)
}

func (e *Engine) NakshatraForDate(utc UtcTime, cfg SankrantiConfig) (PanchangNakshatraInfo, error) {
	out, st := cabi.NakshatraForDate(e.h, utc, cfg)
	return out, statusErr("nakshatra_for_date", st)
}

func (e *Engine) VaarForDate(ep *EOP, utc UtcTime, loc GeoLocation, cfg RiseSetConfig) (VaarInfo, error) {
	out, st := cabi.VaarForDate(e.h, ep.h, utc, loc, cfg)
	return out, statusErr("vaar_for_date", st)
}

func (e *Engine) HoraForDate(ep *EOP, utc UtcTime, loc GeoLocation, cfg RiseSetConfig) (HoraInfo, error) {
	out, st := cabi.HoraForDate(e.h, ep.h, utc, loc, cfg)
	return out, statusErr("hora_for_date", st)
}

func (e *Engine) GhatikaForDate(ep *EOP, utc UtcTime, loc GeoLocation, cfg RiseSetConfig) (GhatikaInfo, error) {
	out, st := cabi.GhatikaForDate(e.h, ep.h, utc, loc, cfg)
	return out, statusErr("ghatika_for_date", st)
}

func (e *Engine) MasaForDate(utc UtcTime, cfg SankrantiConfig) (MasaInfo, error) {
	out, st := cabi.MasaForDate(e.h, utc, cfg)
	return out, statusErr("masa_for_date", st)
}

func (e *Engine) AyanaForDate(utc UtcTime, cfg SankrantiConfig) (AyanaInfo, error) {
	out, st := cabi.AyanaForDate(e.h, utc, cfg)
	return out, statusErr("ayana_for_date", st)
}

func (e *Engine) VarshaForDate(utc UtcTime, cfg SankrantiConfig) (VarshaInfo, error) {
	out, st := cabi.VarshaForDate(e.h, utc, cfg)
	return out, statusErr("varsha_for_date", st)
}

func (e *Engine) PanchangComputeEx(ep *EOP, lsk *LSK, req PanchangComputeRequest) (PanchangOperationResult, error) {
	out, st := cabi.PanchangComputeEx(e.h, ep.h, lsk.h, req)
	return out, statusErr("panchang_compute_ex", st)
}
